src/engine/strat_engine/liminal.rs

```diff
@@ -629,24 +629,25 @@ impl LiminalDevices {
         Ok(unlocked)
     }
 
-    /// Get a list of pool UUIDs from all of the LUKS2 devices that are currently
-    /// locked in the set of pools that are not yet set up.
-    pub fn locked_pool_uuids(&self) -> Vec<PoolUuid> {
+    /// Get a mapping of pool UUIDs from all of the LUKS2 devices that are currently
+    /// locked to their key descriptions in the set of pools that are not yet set up.
+    // Precondition: All devices for a given errored pool have been determined to have
+    // the same key description.
+    pub fn locked_pools(&self) -> HashMap<PoolUuid, KeyDescription> {
         self.errored_pool_devices
             .iter()
             .filter_map(|(pool_uuid, map)| {
-                let has_locked = map.iter().any(|(_, info)| {
-                    if let LInfo::Luks(_) = info {
-                        true
+                map.iter().next().and_then(|(_, info)| {
```

Comment:
Based on our discussion yesterday where we talked about the precondition, I'm not sure what you mean by the devices can have different key descriptions if they're open. We never allow different key descriptions in the LUKS2 metadata whether the devices are open or not if the encrypted devices are created through stratisd so I'll need some clarification around why the precondition doesn't hold. If the encrypted device is created outside of stratisd, it will not be registered as any kind of Stratis device until it has been unlocked and once it is unlocked, it would be registered as an unencrypted device simply because we are not managing the LUKS setup in that case. Based on this, I'm under the impression that the precondition does hold and that we can simply check the first device.

Reply:
At present, the multiple key descriptions for the same pool UUUID could not be accomplished w/out manually editing the LUKS metadata. But @bgurney-rh edits metadata all the time, so it is doable.

My concern is really consistency w/in the code base. Our code should not implicitly assert that this is possible and implicitly assert that it is impossible at the same time. But I'm willing to remove the existing check in setup_pool for this unlikely event, and that will make me happier w/ not checking for it here.
See: https://github.com/stratis-storage/stratisd/pull/2182, commit https://github.com/stratis-storage/stratisd/pull/2182/commits/92c3d9ef4bce0be6adc3781833e7c8c4fdcb6b41.

Please go ahead and resolve the first concern, about the possibility that _some_ but not all of the devices belonging to the pool may be locked, which is easy to achieve by unlocking and then bringing another device online, but also feel free to assume that there is only one key description among all the devices. I'll put https://github.com/stratis-storage/stratisd/pull/2182 up for review.